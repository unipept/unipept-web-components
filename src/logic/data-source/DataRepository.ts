import Sample from '../data-management/Sample';
import ProgressListener from '../patterns/ProgressListener';
import ProgressPublisher from '../patterns/ProgressPublisher';
import MPAConfig from '../data-management/MPAConfig';

import TaxaDataSource from './TaxaDataSource';
import GoDataSource from './GoDataSource';
import EcDataSource from './EcDataSource';

import { PeptideContainerProcessor } from '../processors/peptide/container/PeptideContainerProcessor';
import { ProcessedPeptideContainer } from '../data-management/ProcessedPeptideContainer';
import { TaxaPeptideProcessor } from '../processors/peptide/TaxaPeptideProcessor';
import { GOPeptideProcessor } from '../processors/peptide/GOPeptideProcessor';
import { ECPeptideProcessor } from '../processors/peptide/ECPeptideProcessor';
import { Ontologies } from "../data-management/ontology/Ontologies";
import { PeptideData } from '../api/pept2data/Response';

export default class DataRepository extends ProgressPublisher implements ProgressListener
{
    private readonly _sample: Sample;

    private _processor: PeptideContainerProcessor;
    private _processedPeptideContainer: Promise<ProcessedPeptideContainer>;

    private _mpaConfig: MPAConfig;

    private _taxaSourceCache: TaxaDataSource;
    private _goSourceCache: GoDataSource;
    private _ecSourceCache: EcDataSource;

    public constructor(sample: Sample, mpaConfig: MPAConfig) 
    {
        super();

        this._processor = new PeptideContainerProcessor();
        this._processor.registerProgressListener(this);

        this._sample = sample;
        this._mpaConfig = mpaConfig;
    }

    public async createTaxaDataSource(): Promise<TaxaDataSource> {
        if (!this._taxaSourceCache) 
        {
            let processedPeptideContainer = await this._processedPeptideContainer;
            // set lineages already calculated in the global NCBITaxonomy Ontology
            processedPeptideContainer.response.response.forEach((data: PeptideData) => {
                Ontologies.ncbiTaxonomy.setLineage(data.lca, data.lineage);
            });

            this._taxaSourceCache = new TaxaDataSource(
                TaxaPeptideProcessor.process(processedPeptideContainer), this);
        }
        return this._taxaSourceCache;
    }

    /**
     * Creates a new GoDataSource, or returns one from the cache if the requested namespace was already queried.
     */
    public async createGoDataSource(): Promise<GoDataSource> {
        if (!this._goSourceCache) 
        {
            let processedPeptideContainer = await this._processedPeptideContainer;
            this._goSourceCache = new GoDataSource(
                GOPeptideProcessor.process(processedPeptideContainer),
                processedPeptideContainer, 
                this
            );
        }
        return this._goSourceCache;
    }

    public async createEcDataSource(): Promise<EcDataSource> {
        if (!this._ecSourceCache) 
        {
            let processedPeptideContainer = await this._processedPeptideContainer;
            this._ecSourceCache = new EcDataSource(
                ECPeptideProcessor.process(processedPeptideContainer),
                processedPeptideContainer,
                this
            );
        }
        return this._ecSourceCache;
    }
    
    public async getProcessedPeptideContainer() : Promise<ProcessedPeptideContainer> 
    {
        if (!this._processedPeptideContainer) {
            this._processedPeptideContainer = this._processor.process(this._sample.peptideContainer, this._mpaConfig);
        }
        return this._processedPeptideContainer;
    }

    onProgressUpdate(progress: number): void 
    {
        this.updateProgress(progress)
    }
}
