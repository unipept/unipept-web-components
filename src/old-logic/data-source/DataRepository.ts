import TaxaDataSource from "./TaxaDataSource";
import GoDataSource from "./GoDataSource";
import EcDataSource from "./EcDataSource";
import InterproDataSource from "./InterproDataSource";

export default abstract class DataRepository {
    protected _taxaSourceCache: TaxaDataSource;
    protected _goSourceCache: GoDataSource;
    protected _ecSourceCache: EcDataSource;
    protected _interproSourceCache: InterproDataSource;

    public async createTaxaDataSource(): Promise<TaxaDataSource> {
        if (!this._taxaSourceCache) {
            await this.initTaxaDataSource();
        }
        return this._taxaSourceCache;
    }

    public async createGoDataSource(): Promise<GoDataSource> {
        if (!this._goSourceCache) {
            await this.initGoDataSource();
        }
        return this._goSourceCache;
    }

    public async createEcDataSource(): Promise<EcDataSource> {
        if (!this._ecSourceCache) {
            await this.initEcDataSource();
        }
        return this._ecSourceCache;
    }

    public async createInterproDataSource(): Promise<InterproDataSource> {
        if (!this._interproSourceCache) {
            await this.initInterproDataSource();
        }
        return this._interproSourceCache;
    }

    protected abstract async initTaxaDataSource(): Promise<void>;
    protected abstract async initGoDataSource(): Promise<void>;
    protected abstract async initEcDataSource(): Promise<void>;
    protected abstract async initInterproDataSource(): Promise<void>;
}
