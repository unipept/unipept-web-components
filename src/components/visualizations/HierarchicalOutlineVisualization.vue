<template>
    <div>
        <v-text-field style="margin-bottom: -26px;" outlined single-line label="Search for an organism" append-icon="mdi-magnify" v-model="searchTerm"></v-text-field>
        <div id="searchtree" class="treeView multi"></div>
        <div id="tree_data">
            <p>
                Click on a node in the tree to see the peptides associated with that organism.
                Double-click to focus on it.
            </p>
        </div>
    </div>
</template>

<script lang="ts">
    import Vue from "vue";
    import Component from "vue-class-component";
    import {Prop, Watch} from "vue-property-decorator";
    import Assay from "../../logic/data-management/assay/Assay";
    import Tree from "../../logic/data-management/Tree";
    import {constructSearchtree} from "../../logic/data-management/searchtree";
    import TaxaDataSource from "../../logic/data-source/TaxaDataSource";
    import DataRepository from '../../logic/data-source/DataRepository';

    @Component({
        components: {},
        computed: {
            searchTerm: {
                get() {
                    let term = this.$store.getters.selectedTerm;
                    if (term === 'Organism') {
                        return '';
                    }
                    return term;
                },
                set(val) {
                    // should do nothing!
                }
            },
            watchableSelectedSearchTerm: {
                get(): string {
                    return this.$store.getters.selectedTerm
                }
            }
        }
    })
    export default class HierarchicalOutlineVisualization extends Vue {
        @Prop({required: true}) dataRepository: DataRepository;

        searchTree!: any;

        mounted() {
            this.initSearchTree();
        }

        @Watch('dataset') onDatasetChanged() {
            this.initSearchTree();
        }

        @Watch('watchableSelectedS') onActiveSearchTermChanged(newSearchTerm: string, oldSearchTerm: string) {
            if (this.searchTree && newSearchTerm !== "") {
                setTimeout(() => {
                    this.searchTree.search(newSearchTerm);
                }, 500);
            }
        }

        @Watch('watchableSelectedSearchTerm')
        private onSelectedSearchTermChanged(newSearchTerm: string) {
            if(this.searchTree && newSearchTerm) {
                this.searchTree.search(newSearchTerm);
            }
        }

        private async initSearchTree() {
            if (this.dataRepository != null) {
                let taxaDataSource: TaxaDataSource = await this.dataRepository.createTaxaDataSource();
                let tree: Tree = await taxaDataSource.getTree();
                this.searchTree = constructSearchtree(tree, this.$store.getters.searchSettings.il, () => {});
            }
        }

        get activeSearchTerm() {
            return this.$store.getters.selectedTerm;
        }
    }
</script>

<style lang="less">
    @import './../../assets/style/variables.css.less';

    .treeView {
        -moz-user-select: none;
        position: relative;
        margin-top: 5px;
        float: left;
    }

    .treeView ul {
        margin: -5px 0 0 -1.5em;
        padding: 0 0 0 1.5em;
    }

    .treeView ul ul {
        background: url(./../../assets/images/hierarchical_outline/list-item-contents.png) repeat-y left;
    }

    .treeView li:last-child > ul {
        background-image: none;
    }

    .treeView li {
        margin: 0;
        padding: 5px 0 0;
        background: url(./../../assets/images/hierarchical_outline/list-item-root.png) no-repeat top left;
        list-style-position: inside;
        list-style-image: url(./../../assets/images/hierarchical_outline/button.png);
        cursor: auto;
        white-space: nowrap;
        text-overflow: ellipsis;
    }

    .treeView li.collapsibleListOpen {
        list-style-image: url(./../../assets/images/hierarchical_outline/button-open.png);
        cursor: pointer;
    }

    .treeView li.collapsibleListClosed {
        list-style-image: url(./../../assets/images/hierarchical_outline/button-closed.png);
        cursor: pointer;
    }

    .treeView li.collapsibleListClosed ul {
        display: none;
    }

    .treeView li li {
        background-image: url(./../../assets/images/hierarchical_outline/list-item.png);
        padding-left: 1.5em;
    }

    .treeView li:last-child {
        background-image: url(./../../assets/images/hierarchical_outline/list-item-last.png);
    }

    .treeView li.root {
        background-image: none;
        list-style-type: none;
        list-style-image: none;
    }

    .treeView li.unmatch {
        display: none;
    }

    .treeView li.leaf {
        background-image: url(./../../assets/images/hierarchical_outline/list-item-leaf.png);
        list-style-type: none;
        list-style-image: none;
    }

    .treeView li.leaf:last-child {
        background-image: url(./../../assets/images/hierarchical_outline/list-item-last-leaf.png);
    }

    .treeView span.clicked {
        background: @gray-lighter !important;
        border: 1px solid #ccc !important;
        border-radius: 2px;
    }

    .treeView li.leaf span {
        cursor: pointer;
    }

    #tree_data {
        float: right;
        font-family: Roboto,'Helvetica Neue',Helvetica,Arial,sans-serif;
        font-size: 12px;
        color: rgb(85, 85, 85);
    }

    #searchtree {
        margin-bottom: 16px;
    }
</style>
