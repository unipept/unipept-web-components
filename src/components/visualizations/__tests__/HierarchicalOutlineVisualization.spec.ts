import { mount, createLocalVue } from "@vue/test-utils"
import Vue from "vue"
import Vuetify from "vuetify"
import Mock from "../../../test/Mock"
import DataRepository from "../../../logic/data-source/DataRepository";
import HierarchicalOutlineVisualization from "./../HierarchicalOutlineVisualization.vue";
import "jsdom-worker-fix";
import Vuex from "vuex";

Vue.use(Vuetify);
Vue.use(Vuex);

const localVue = createLocalVue();

describe("HierarchicalOutlineVisualization", () => {
    let vuetify;
    let getters;
    let store;

    beforeEach(() => {
        vuetify = new Vuetify();

        getters = {
            selectedTerm: () => "Organism",
            searchSettings: () => {
                return {
                    il: true,
                    dupes: true,
                    missed: false
                }
            }
        }

        store = new Vuex.Store({
            getters
        });
    });

    it("renders simple datasets", (done) => {
        let mock: Mock = new Mock();
        mock.mockDataRepository().then((dataRepository: DataRepository) => {
            const wrapper = mount(HierarchicalOutlineVisualization, {
                store,
                localVue,
                vuetify,
                propsData: {
                    dataRepository: dataRepository
                }
            });
    
            // Wait for all async operations to be finished, before expecting anything.
            const flushPromises = () => new Promise(setImmediate);
            flushPromises().then(() => {
                expect(wrapper.html()).toMatchSnapshot();
                // Call done here in the promise resolve.
                done();
                
            })
        }).catch(err => console.log(err));
    });
});
