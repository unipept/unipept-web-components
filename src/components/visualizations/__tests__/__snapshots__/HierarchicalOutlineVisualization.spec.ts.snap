// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HierarchicalOutlineVisualization renders simple datasets 1`] = `
"<div><div class=\\"v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--single-line v-text-field--enclosed v-text-field--outlined\\" style=\\"margin-bottom: -26px;\\"><div class=\\"v-input__control\\"><div class=\\"v-input__slot\\"><fieldset aria-hidden=\\"true\\"><legend><span>â€‹</span></legend></fieldset><div class=\\"v-text-field__slot\\"><input name=\\"tree_search\\" id=\\"tree_search\\" type=\\"text\\"></div><div class=\\"v-input__append-inner\\"><div class=\\"v-input__icon v-input__icon--append\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-magnify theme--light\\"></i></div></div></div><div class=\\"v-text-field__details\\"><div class=\\"v-messages theme--light\\"><span name=\\"message-transition\\" tag=\\"div\\" class=\\"v-messages__wrapper\\"></span></div></div></div></div> <div id=\\"searchtree\\" class=\\"treeView multi\\"></div> <div id=\\"tree_data\\"><p>
            Click on a node in the tree to see the peptides associated with that organism.
            Double-click to focus on it.
        </p></div></div>"
`;
